pkg/retriever/graphsyncretriever.go

```diff
@@ -148,12 +166,32 @@ func (r *graphsyncRetrieval) RetrieveFromAsyncCandidates(asyncCandidates types.I
 				return
 			}
 			for _, candidate := range candidates {
-				candidate := candidate
-				waitGroup.Add(1)
-				go func() {
-					defer waitGroup.Done()
-					runRetrievalCandidate(ctx, r.GraphSyncRetriever, r.request, r.Client, r.request.LinkSystem, retrieval, phaseStartTime, candidate)
-				}()
+				// Check if we already started a retrieval for this candidate
+				r.candidateMetdataMut.RLock()
+				currMetadata, seenCandidate := r.candidateMetadata[candidate.MinerPeer.ID]
+				r.candidateMetdataMut.RUnlock()
+				if seenCandidate {
+					// Don't start a new retrieval, but update the metadata if it's more favorable
+					newMetadata := candidate.Metadata.Get(multicodec.TransportGraphsyncFilecoinv1).(*metadata.GraphsyncFilecoinV1)
+					if metadataCompare(newMetadata, currMetadata) {
+						r.candidateMetdataMut.Lock()
+						r.candidateMetadata[candidate.MinerPeer.ID] = newMetadata
+						r.candidateMetdataMut.Unlock()
+					}
+				} else {
```

Comment:
A `return` might make this a tad cleaner by avoiding the `else`; your call tho

Reply:
actually .. now I see that the most important part of this code, the `runRetrievalCandidate` is now nested even further, I'm more strongly of the opinion that we should early `return` and avoid the nesting. The most important end-point of the code shouldn't be hidden so deeply in nesting IMO.

Reply:
thumbs up on return early.

[gaurd clause = easier to understand than if-else](https://deviq.com/design-patterns/guard-clause)

Reply:
ok follow up: I see why there's no return -- cause it's a loop. so we need a `continue` instead.

Reply:
I agree with returning early; I usually try to. I'll see if I can clean this up with a continue.